core/src/Dscp/Core/FairCV.hs

```diff
@@ -0,0 +1,130 @@
+-- Tagged Merkle proofs
+---------------------------------------------------------------------------
+
+-- | Type tag for proofs/CVs not checked yet.
+data Unchecked
+
+-- | Type tag for validated proofs/CVs.
+data Valid
+
+-- | 'MerkleProof' tagged by validation tag. When using this module,
+-- it is impossible to obtain a value of 'TaggedProof Valid a' without
+-- doing a validity check.
+newtype TaggedProof v a = TaggedProof
+    { unTaggedProof :: MerkleProof a
+    } deriving (Show, Eq, Generic)
+
+-- | Safe constructor for 'TaggedProof', which can yield only 'Unckecked'
+-- proof.
+mkTagged :: MerkleProof a -> TaggedProof Unchecked a
+mkTagged = TaggedProof
+
+-- | If 'Unchecked' proof is valid, make it 'Valid', otherwise return an
+-- error.
+validateTaggedProof
+    :: HasHash a
+    => TaggedProof Unchecked a
+    -> Either Text (TaggedProof Valid a)
+validateTaggedProof (TaggedProof unchecked) =
+    if isJust (computeMerkleRoot unchecked)
+    then Right $ TaggedProof unchecked
+    else Left "Proof validation has failed"
+
+-- | Merges two already validated Merkle proofs, checking that all common
+-- inner nodes match.
+mergeProofs
+    :: TaggedProof Valid a
+    -> TaggedProof Valid a
+    -> Either Text (TaggedProof Valid a)
+mergeProofs (TaggedProof a) (TaggedProof b)
+    | pnSig a /= pnSig b = Left "Node signatures mismatch"
+    | otherwise = TaggedProof <$> merge a b
+  where
+    merge (ProofBranch s l r) (ProofBranch _ l' r') =
+        let mergeUntag t t' =
+                unTaggedProof <$>
+                mergeProofs (TaggedProof t) (TaggedProof t')
+        in ProofBranch s <$> mergeUntag l l' <*> mergeUntag r r'
+    merge ProofBranch{} ProofLeaf{} =
+        Left "Inner node on the left is mismatched with leaf on the right"
+    merge ProofLeaf{} ProofBranch{} =
+        Left "Leaf on the left is mismatched with inner node on the right"
+    -- No need to compare values in leaves, because proofs are 'Valid',
+    -- and hash function should be collision-resistnant.
+    merge a'@ProofLeaf{} ProofLeaf{} = pure a'
+    merge a' ProofPruned{} = pure a'
+    merge ProofPruned{} b' = pure b'
+
+---------------------------------------------------------------------------
+-- FairCV
+---------------------------------------------------------------------------
+
+-- | FairCV datatype. Proofs are divided by educators (designated by their
+-- addresses) and blocks (designated by their hashes).
+-- Phantom type parameter @v@ distinguishes between already validated
+-- 'FairCV' and 'FairCV' not validated yet.
+newtype FairCV v = FairCV
+    { unFairCV :: Map Address (Map PrivateHeaderHash (TaggedProof v PrivateTx))
+    } deriving (Show, Eq)
+
+-- | If 'Unchecked' FairCV is valid, make it 'Valid', otherwise return an
+-- error.
+validateFairCV :: FairCV Unchecked -> Either Text (FairCV Valid)
+validateFairCV (FairCV cv) =
+    FairCV <$> traverse (traverse validateTaggedProof) cv
```

Comment:
Btw, according to your plan, we probably want it to return the exact list of false headers.